### ISSUE TITLE
Recalculating checksums results in exponential performance degradation

### ISSUE THREAD

**Issue by dhsmith1001** [liquibase/liquibase#6564]
### Search first

- [x] I searched and no similar issues were found

### Description

I have a database that reruns Liquibase frequently to potentially add table partitions as well as execute a procedure that copies data from one table to another. This results in a higher number of rows in the databasechangelog table than "normal" scenarios. In my test case, 27k rows.

I have always cleared checksums before running Liquibase due to the historical inability of Liquibase to accurately calculate checksums (I suspect due to the use of system properties within the changesets).

In this case, it's obvious that Liquibase is recalculating the checksums for the exact same changeset once for every instance of the changeset in the table. For example, there are changesets that have executed 1577 times. While Liquibase updates all 1577 rows at once:

[2024-12-04 16:16:34] FINE [liquibase.changelog] Updating null or out of date checksum on changeSet JdbcAuditChangeLog.xml::430::SYSTEM to correct value
[2024-12-04 16:16:34] FINE [liquibase.executor] UPDATE public.databasechangelog SET MD5SUM = '9:cc72d6f444baf60e74560b719ecc7687' WHERE ID = '430' AND AUTHOR = 'SYSTEM' AND FILENAME = 'JdbcAuditChangeLog.xml'
[2024-12-04 16:16:34] FINE [liquibase.executor] 1577 row(s) affected

It does it 1577 times. Searching my log file for that statement yields a count of 1577:

$ grep "UPDATE public.databasechangelog SET MD5SUM = '9:cc72d6f444baf60e74560b719ecc7687' WHERE ID = '430' AND AUTHOR = 'SYSTEM' AND FILENAME = 'JdbcAuditChangeLog.xml" db-init-audit-log.txt  | wc -l

 **1577**

### Steps To Reproduce

1. Import the attached databasechangelog.csv into an empty databasechangelog table.
2. Run clearchecksums on this database, with --log-level=debug
3. Note how many update statements are logged for clearing the checksums.

### Expected/Desired Behavior

Each clearchecksums update statement should execute only once per changeset ID.

### Liquibase Version

4.30.0

### Liquibase Integration

CLI

### Liquibase Extensions

n/a

### OS and/or Infrastructure Type/Provider

Ubuntu

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by dhsmith1001** [comment]
My changelog export:
[databasechangelog.csv](https://github.com/user-attachments/files/18028308/databasechangelog.csv)

**Comment by tati-qalified** [comment]
Hi @dhsmith1001, thank you for reaching out.
I followed the steps to reproduce, but my logs show only `[2025-01-27 16:42:44] FINE [liquibase.executor] UPDATE lbcat.DATABASECHANGELOG SET MD5SUM = NULL` and the execution takes milliseconds, even with 27000+ entries.

What Liquibase version were you using before you noticed this degradation? Do you see it with 4.31.0 as well?

**Comment by dhsmith1001** [comment]
> Hi [@dhsmith1001](https://github.com/dhsmith1001), thank you for reaching out. I followed the steps to reproduce, but my logs show only `[2025-01-27 16:42:44] FINE [liquibase.executor] UPDATE lbcat.DATABASECHANGELOG SET MD5SUM = NULL` and the execution takes milliseconds, even with 27000+ entries.
> 
> What Liquibase version were you using before you noticed this degradation? Do you see it with 4.31.0 as well?

You're looking at the clearCheckSums log, not the update log after clearing checksums. The title indicates "recalculating checksums" which happens when running update subsequent to clearing checksums.

**Comment by dhsmith1001** [comment]
This zip file contains all the necessary files to reproduce the issue. To use it:

1. Unzip Liquibase6564.zip into some working directory
2. cd to the demo6564 directory
3. Run checksum-defect.sh redirecting stdout and stderr to a file: ./checksum-defect.sh > log.txt 2>&1
4. Search log.txt for an example update statement: grep "UPDATE public.databasechangelog SET MD5SUM = '9:cc72d6f444baf60e74560b719ecc7687' WHERE ID = '430' " log.txt

[Liquibase6564.zip](https://github.com/user-attachments/files/18591080/Liquibase6564.zip)

**Comment by tati-qalified** [comment]
Thank you for the reproducer, I can see the problem now. Our development team will be in contact through your submitted PR.

Thanks again,
Tatiana

**Comment by dhsmith1001** [comment]
After discussion with @MalloD12 I am closing this issue and have created a new issue for the underlying problem of the creation of duplicate rows in databasechangelog: https://github.com/liquibase/liquibase/issues/6821
